Source/BSGConnectivity.m

```diff
@@ -35,85 +84,51 @@
+    }
 }
 
-@interface BSGConnectivity ()
-
-@property(nonatomic, assign) SCNetworkReachabilityRef reachabilityRef;
-@property(nonatomic, strong) dispatch_queue_t serialQueue;
-@property(nonatomic, copy) CallbackBlock callbackBlock;
-
-@end
-
 @implementation BSGConnectivity
 
-- (instancetype)initWithURL:(NSURL *)url
-                changeBlock:(ConnectivityChange)changeBlock {
-    if (self = [super init]) {
-        NSString *hostName = [url host];
-        _reachabilityRef = SCNetworkReachabilityCreateWithName(NULL, [hostName UTF8String]);
-
-        _connectivityChangeBlock = [changeBlock copy];
-        self.serialQueue = dispatch_queue_create("com.bugsnag.cocoa", NULL);
++ (void)monitorURL:(NSURL *)URL usingCallback:(BSGConnectivityChangeBlock)block {
+    static dispatch_once_t once_t;
+    static dispatch_queue_t reachabilityQueue;
+    dispatch_once(&once_t, ^{
+        reachabilityQueue = dispatch_queue_create("com.bugsnag.cocoa.connectivity", DISPATCH_QUEUE_SERIAL);
+    });
 
-        __weak id weakSelf = self;
-        self.callbackBlock = ^(SCNetworkReachabilityFlags flags) {
-          if (weakSelf) {
-              [weakSelf connectivityChanged:flags];
-          }
-        };
+    NSString *host = [URL host];
+    if (![self isValidHostname:host]) {
+        return;
     }
-    return self;
-}
-
-- (void)dealloc {
-    [self stopWatchingConnectivity];
 
-    if (self.reachabilityRef) {
-        CFRelease(self.reachabilityRef);
-        self.reachabilityRef = nil;
+    bsg_reachability_ref = SCNetworkReachabilityCreateWithName(NULL, [host UTF8String]);
+    if (bsg_reachability_ref) { // Can be null if a bad hostname was specified
+        bsg_reachability_change_block = block;
+        SCNetworkReachabilitySetCallback(bsg_reachability_ref,
+                                         BSGConnectivityCallback, NULL);
+        SCNetworkReachabilitySetDispatchQueue(bsg_reachability_ref,
+                                              reachabilityQueue);
     }
 }
 
 /**
- * Sets a callback with SCNetworkReachability
+ * Check if the host is valid and not equivalent to localhost, from which we can
+ * never truly disconnect. 🏡
+ * There are also system handlers for localhost which we don't want to catch
+ * inadvertently.
  */
-- (void)startWatchingConnectivity {
-    SCNetworkReachabilityContext context = {
-        .version = 0,
-        .info = (__bridge void *)self.callbackBlock,
-        .retain = CFRetain,
-        .release = CFRelease,
-    };
-
-    if (self.reachabilityRef) {
-        SCNetworkReachabilitySetCallback(self.reachabilityRef,
-                                         BSGConnectivityCallback, &context);
-        SCNetworkReachabilitySetDispatchQueue(self.reachabilityRef,
-                                              self.serialQueue);
-    }
++ (BOOL)isValidHostname:(NSString *)host {
+    return host.length > 0 && ![host isEqualToString:@"localhost"];
```

Comment:
This doesn't catch `127.0.0.1`.  Since there's a (potential) loss of network at this point is an explicit check for IP address better than a more general host to IP resolution?  On desktop platforms this will also fail to catch `/etc/hosts` host aliasing but since this (along with endpoint reconfiguration) is an intentional user-action it feels safe to ignore the edge case.